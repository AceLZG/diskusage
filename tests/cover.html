
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">github.com/aleksaan/diskusage/diskusage/common.go (84.8%)</option>
				
				<option value="file1">github.com/aleksaan/diskusage/diskusage/input_args.go (0.0%)</option>
				
				<option value="file2">github.com/aleksaan/diskusage/diskusage/types.go (0.0%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">package diskusage

import (
        "io/ioutil"
        "math"
        "os"
)

//pairs of key and scale power x, when 1024^x is scale of size
var sizeUnits = map[string]float64{
        "b":  1,
        "Kb": 2,
        "Mb": 3,
        "Gb": 4,
        "Tb": 5,
        "Pb": 6,
}

var sortedKeysSizeUnits = []string{"b", "Kb", "Mb", "Gb", "Tb", "Pb"}

//-----------------------------------------------------------------------------------------

//ScanDir - scan directory and return its size
func ScanDir(files *TFiles, path string, depth int) int64 <span class="cov8" title="1">{
        //read content of folder
        osfiles, _ := ioutil.ReadDir(path)

        var dirsize int64

        //calc total size throught folder content
        for _, osfile := range osfiles </span><span class="cov8" title="1">{
                file := ScanFile(path+"/"+osfile.Name(), depth)
                if file.IsDir </span><span class="cov0" title="0">{
                        file.Size = ScanDir(files, file.Path, depth+1)
                }</span>
                <span class="cov8" title="1">file.SetAdaptedSizeOfFile(&amp;InputArgs)
                dirsize += file.Size
                *files = append(*files, *file)</span>
        }

        <span class="cov8" title="1">return dirsize</span>
}

//-----------------------------------------------------------------------------------------

//ScanFile - scan dir/file parameters
func ScanFile(path string, depth int) *TFile <span class="cov8" title="1">{
        f := &amp;TFile{}
        f.Path = path
        f.Depth = depth

        //if file or folder is not accessible then return nil
        dir, err := os.Stat(path)
        if err != nil </span><span class="cov0" title="0">{
                f.IsNotAccessible = true
                f.IsNotAccessibleMessage = err.Error()
        }</span>

        <span class="cov8" title="1">f.IsDir = dir.IsDir()

        if !dir.IsDir() </span><span class="cov8" title="1">{
                f.Size = dir.Size()
        }</span>

        <span class="cov8" title="1">return f</span>
}

//SetAdaptedSizeOfFile - set file size adapted to InputArgs.FixUnit units or to a flexible useful units
func (f *TFile) SetAdaptedSizeOfFile(inputargs *TInputArgs) <span class="cov8" title="1">{

        var size = float64(f.Size)
        var unit string
        var power float64

        if len(inputargs.FixUnit) &gt; 0 </span><span class="cov0" title="0">{
                unit = inputargs.FixUnit
                power = sizeUnits[inputargs.FixUnit]
        }</span> else<span class="cov8" title="1"> {
                for _, unit = range sortedKeysSizeUnits </span><span class="cov8" title="1">{
                        power = sizeUnits[unit]
                        if size &lt; math.Pow(1024, power) </span><span class="cov8" title="1">{
                                break</span>
                        }
                }
        }
        <span class="cov8" title="1">f.AdaptedUnit = unit
        f.AdaptedSize = size / math.Pow(1024, power-1)</span>
}
</pre>
		
		<pre class="file" id="file1" style="display: none">package diskusage

import (
        "errors"
        "flag"
        "fmt"
        "strconv"
        "strings"
)

//LimitDefault - default Limit value
const LimitDefault = 10 //default value for a arglimit
//DepthDefault - default Depth value
const DepthDefault = 1 //default depth in results

//InputArgs - input arguments
var InputArgs TInputArgs

//TInputArgs - the programm arguments
type TInputArgs struct {
        Paths   string //list of folders to analyze separated by ;
        Limit   int    //limit folders in results
        Depth   int    //depth of subfolders in results (-1 - all, 1 - only current, 2 and more - 2 and more)
        FixUnit string //fixed size unit in a results presetation (b, Kb, Mb, ...). Has a upper priority than "argmaxunit". Must be in sizeUnits.
}

//SetPaths - init Paths field
func (t *TInputArgs) SetPaths(paths *string) error <span class="cov0" title="0">{
        *paths = strconv.Quote(*paths)
        *paths = strings.Replace(*paths, "\\", "/", -1)
        *paths, _ = strconv.Unquote(*paths)
        if len(*paths) == 0 </span><span class="cov0" title="0">{
                return errors.New("Error! Argument 'path' could not be an empty string")
        }</span>
        <span class="cov0" title="0">t.Paths = *paths
        return nil</span>
}

//SetLimit - init Limit field
func (t *TInputArgs) SetLimit(limit *int) error <span class="cov0" title="0">{
        if *limit &lt; 1 </span><span class="cov0" title="0">{
                fmt.Printf("Argument 'limit' is negative (%d) and has been set to default value (%d)", *limit, LimitDefault)
                *limit = LimitDefault //set to default value
        }</span>
        <span class="cov0" title="0">t.Limit = *limit
        return nil</span>
}

//SetFixUnit - init FixUnit field
func (t *TInputArgs) SetFixUnit(fixunit *string) error <span class="cov0" title="0">{
        if _, ok := sizeUnits[*fixunit]; !ok &amp;&amp; len(*fixunit) &gt; 0 </span><span class="cov0" title="0">{
                return errors.New("Error! Argument 'fixunit' is not in allowable range {b, Kb, Mb, Gb, Tb, Pb}")
        }</span>
        <span class="cov0" title="0">if len(*fixunit) &gt; 0 </span><span class="cov0" title="0">{
                fmt.Printf("Results will be represented with fixed units style in '%s'\n", *fixunit)
        }</span>
        <span class="cov0" title="0">return nil</span>
}

//SetDepth - init Depth field
func (t *TInputArgs) SetDepth(depth *int) error <span class="cov0" title="0">{
        if *depth &lt; 0 </span><span class="cov0" title="0">{
                fmt.Printf("Argument 'depth' is negative (%d) and has been set to default value (%d)", *depth, DepthDefault)
        }</span>
        <span class="cov0" title="0">return nil</span>
}

//PrintArguments - print arguments
func (t TInputArgs) PrintArguments() error <span class="cov0" title="0">{
        fmt.Println("Arguments:")
        fmt.Printf("   path: %q\n", t.Paths)
        fmt.Printf("   limit: %d\n", t.Limit)
        fmt.Printf("   fixunit: %s\n", t.FixUnit)
        fmt.Printf("   depth: %d\n", t.Depth)
        return nil
}</span>

//ParseCLIArguments - parses input arguments
func (t *TInputArgs) ParseCLIArguments() <span class="cov0" title="0">{

        var argpaths = flag.String("path", "", "Folders paths separated by semicolon (required)")
        var arglimit = flag.Int("limit", LimitDefault, fmt.Sprintf("Number of folders printed in a results (%d by default)", LimitDefault))
        var argfixunit = flag.String("fixunit", "", "Fixed size unit for a results represetation. Should be one of {b, Kb, Mb, Gb, Tb, Pb}.")
        var argdepth = flag.Int("depth", DepthDefault, "Depth of subfolders.")

        //parse argument
        flag.Parse()

        //processing arguments
        t.SetPaths(argpaths)
        //checkError(err)

        t.SetLimit(arglimit)
        //checkError(err)

        t.SetFixUnit(argfixunit)
        //checkError(err)

        t.SetDepth(argdepth)
        //checkError(err)
}</span>
</pre>
		
		<pre class="file" id="file2" style="display: none">package diskusage

import (
        "fmt"
        "math"
        "sort"
)

//TFile - struct for file object
type TFile struct {
        Path                   string
        Size                   int64
        IsDir                  bool
        Depth                  int
        IsNotAccessible        bool
        IsNotAccessibleMessage string
        AdaptedSize            float64
        AdaptedUnit            string
}

//TFiles - struct for files array object
type TFiles []TFile

type sizeSorter []TFile

func (a sizeSorter) Len() int           <span class="cov0" title="0">{ return len(a) }</span>
func (a sizeSorter) Swap(i, j int)      <span class="cov0" title="0">{ a[i], a[j] = a[j], a[i] }</span>
func (a sizeSorter) Less(i, j int) bool <span class="cov0" title="0">{ return a[i].Size &lt; a[j].Size }</span>

type sizeDescSorter []TFile

func (a sizeDescSorter) Len() int           <span class="cov0" title="0">{ return len(a) }</span>
func (a sizeDescSorter) Swap(i, j int)      <span class="cov0" title="0">{ a[i], a[j] = a[j], a[i] }</span>
func (a sizeDescSorter) Less(i, j int) bool <span class="cov0" title="0">{ return a[i].Size &gt; a[j].Size }</span>

//-----------------------------------------------------------------------------------------

//Sort - sort files by size
func (files *TFiles) Sort(by string, order string) <span class="cov0" title="0">{
        switch </span>{
        case by == "size" &amp;&amp; order == "desc":<span class="cov0" title="0">
                sort.Sort(sizeDescSorter(*files))</span>
        case by == "size" &amp;&amp; order == "":<span class="cov0" title="0">
                sort.Sort(sizeSorter(*files))</span>
        default:<span class="cov0" title="0"></span>
        }
}

//-----------------------------------------------------------------------------------------

//CalculateMaxLenFilename - calculating max length filename in files
func (files *TFiles) CalculateMaxLenFilename() int <span class="cov0" title="0">{
        var maxlen = 0
        var c = 0
        for _, f := range *files </span><span class="cov0" title="0">{
                if f.Depth &lt;= InputArgs.Depth </span><span class="cov0" title="0">{
                        c++
                        maxlen = int(math.Max(float64(maxlen), float64(len(f.Path))))
                        //break if we up to defined limit
                        if c+1 &gt;= InputArgs.Limit-1 </span><span class="cov0" title="0">{
                                break</span>
                        }
                }
        }
        <span class="cov0" title="0">return maxlen</span>
}

//-----------------------------------------------------------------------------------------

//PrintFilesSizes - print out sizes of files
func (files *TFiles) PrintFilesSizes() <span class="cov0" title="0">{
        maxlen := files.CalculateMaxLenFilename()

        //print results
        var strfmt = "%3d.| DIR: %-" + fmt.Sprintf("%d", maxlen+2) + "s | SIZE: %6.2f %-4s | DEPTH: %d \n"
        var c = 0
        for _, f := range *files </span><span class="cov0" title="0">{
                if f.Depth &lt;= InputArgs.Depth </span><span class="cov0" title="0">{
                        c++
                        fmt.Printf(strfmt, c, f.Path, f.AdaptedSize, f.AdaptedUnit, f.Depth)

                        //break if we up to defined limit
                        if c+1 &gt;= InputArgs.Limit-1 </span><span class="cov0" title="0">{
                                break</span>
                        }
                }
        }
}
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
