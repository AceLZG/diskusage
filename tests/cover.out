mode: set
github.com/aleksaan/diskusage/diskusage/common.go:24.59,31.33 3 1
github.com/aleksaan/diskusage/diskusage/common.go:41.2,41.16 1 1
github.com/aleksaan/diskusage/diskusage/common.go:31.33,33.17 2 1
github.com/aleksaan/diskusage/diskusage/common.go:36.3,38.33 3 1
github.com/aleksaan/diskusage/diskusage/common.go:33.17,35.4 1 0
github.com/aleksaan/diskusage/diskusage/common.go:47.46,54.16 5 1
github.com/aleksaan/diskusage/diskusage/common.go:59.2,61.18 2 1
github.com/aleksaan/diskusage/diskusage/common.go:65.2,65.10 1 1
github.com/aleksaan/diskusage/diskusage/common.go:54.16,57.3 2 0
github.com/aleksaan/diskusage/diskusage/common.go:61.18,63.3 1 1
github.com/aleksaan/diskusage/diskusage/common.go:69.61,75.32 4 1
github.com/aleksaan/diskusage/diskusage/common.go:86.2,87.48 2 1
github.com/aleksaan/diskusage/diskusage/common.go:75.32,78.3 2 0
github.com/aleksaan/diskusage/diskusage/common.go:78.8,79.43 1 1
github.com/aleksaan/diskusage/diskusage/common.go:79.43,81.36 2 1
github.com/aleksaan/diskusage/diskusage/common.go:81.36,82.10 1 1
github.com/aleksaan/diskusage/diskusage/input_args.go:28.52,32.22 4 0
github.com/aleksaan/diskusage/diskusage/input_args.go:35.2,36.12 2 0
github.com/aleksaan/diskusage/diskusage/input_args.go:32.22,34.3 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:40.49,41.16 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:45.2,46.12 2 0
github.com/aleksaan/diskusage/diskusage/input_args.go:41.16,44.3 2 0
github.com/aleksaan/diskusage/diskusage/input_args.go:50.56,51.60 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:54.2,54.23 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:57.2,57.12 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:51.60,53.3 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:54.23,56.3 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:61.49,62.16 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:65.2,65.12 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:62.16,64.3 1 0
github.com/aleksaan/diskusage/diskusage/input_args.go:69.44,76.2 6 0
github.com/aleksaan/diskusage/diskusage/input_args.go:79.42,101.2 9 0
github.com/aleksaan/diskusage/diskusage/types.go:26.41,26.58 1 0
github.com/aleksaan/diskusage/diskusage/types.go:27.41,27.68 1 0
github.com/aleksaan/diskusage/diskusage/types.go:28.41,28.73 1 0
github.com/aleksaan/diskusage/diskusage/types.go:32.45,32.62 1 0
github.com/aleksaan/diskusage/diskusage/types.go:33.45,33.72 1 0
github.com/aleksaan/diskusage/diskusage/types.go:34.45,34.77 1 0
github.com/aleksaan/diskusage/diskusage/types.go:39.52,40.9 1 0
github.com/aleksaan/diskusage/diskusage/types.go:41.39,42.36 1 0
github.com/aleksaan/diskusage/diskusage/types.go:43.35,44.32 1 0
github.com/aleksaan/diskusage/diskusage/types.go:45.10,45.10 0 0
github.com/aleksaan/diskusage/diskusage/types.go:52.52,55.27 3 0
github.com/aleksaan/diskusage/diskusage/types.go:65.2,65.15 1 0
github.com/aleksaan/diskusage/diskusage/types.go:55.27,56.33 1 0
github.com/aleksaan/diskusage/diskusage/types.go:56.33,60.32 3 0
github.com/aleksaan/diskusage/diskusage/types.go:60.32,61.10 1 0
github.com/aleksaan/diskusage/diskusage/types.go:71.40,77.27 4 0
github.com/aleksaan/diskusage/diskusage/types.go:77.27,78.33 1 0
github.com/aleksaan/diskusage/diskusage/types.go:78.33,83.32 3 0
github.com/aleksaan/diskusage/diskusage/types.go:83.32,84.10 1 0
